x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/registry/ModelRegistry.java

```diff
@@ -662,163 +661,261 @@ public void storeModel(Model model, ActionListener<Boolean> listener, TimeValue
 
-            @Override
-            public void onFailure(Exception exc) {
-                logger.warn(
-                    format("Failed to add inference endpoint [%s] minimal service settings to cluster state", inferenceEntityId),
-                    exc
-                );
-                deleteModel(inferenceEntityId, ActionListener.running(() -> {
-                    listener.onFailure(
-                        new ElasticsearchStatusException(
-                            format(
-                                "Failed to add the inference endpoint [%s]. The service may be in an "
-                                    + "inconsistent state. Please try deleting and re-adding the endpoint.",
-                                inferenceEntityId
+    private record ResponseInfo(List<StoreResponseWithIndexInfo> responses, List<Model> successfullyStoredModels) {}
+
+    private static ResponseInfo getResponseInfo(
+        BulkResponse bulkResponse,
+        Map<String, String> docIdToInferenceId,
+        Map<String, Model> inferenceIdToModel
+    ) {
+        var responses = new ArrayList<StoreResponseWithIndexInfo>();
+        var successfullyStoredModels = new ArrayList<Model>();
+
+        var bulkItems = bulkResponse.getItems();
+        for (int i = 0; i < bulkItems.length; i += 2) {
+            var configurationItem = bulkItems[i];
+            var configStoreResponse = createModelStoreResponse(configurationItem, docIdToInferenceId);
+            var modelFromBulkItem = getModelFromMap(docIdToInferenceId.get(configurationItem.getId()), inferenceIdToModel);
+
+            if (i + 1 >= bulkResponse.getItems().length) {
+                logger.error("Expected an even number of bulk response items, got [{}]", bulkResponse.getItems().length);
+
+                if (configStoreResponse.failed()) {
+                    responses.add(new StoreResponseWithIndexInfo(configStoreResponse, false));
+                } else {
+                    // if we didn't get the last item for some reason, assume it is a failure
+                    responses.add(
+                        new StoreResponseWithIndexInfo(
+                            new ModelStoreResponse(
+                                configStoreResponse.inferenceId(),
+                                RestStatus.INTERNAL_SERVER_ERROR,
+                                new IllegalStateException("Failed to receive part of bulk response")
                             ),
-                            RestStatus.INTERNAL_SERVER_ERROR,
-                            exc
+                            true
                         )
                     );
-                }));
+                }
+                return new ResponseInfo(responses, successfullyStoredModels);
             }
-        };
+
+            var secretsItem = bulkItems[i + 1];
+            var secretsStoreResponse = createModelStoreResponse(secretsItem, docIdToInferenceId);
```

Comment:
Is it worth doing some kind of check that the inference ID for the secrets item matches the inference ID for the configuration item? Or is it not possible to lose items from part way through the bulk response, only at the end?